# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/webrender_bindings/DCLayerTree.cpp
# Commit: f3b490c076d6
# Full Hash: f3b490c076d623d6ef3bcfd297a0522d9727477c
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2020-01-23 21:42:24
# Regressor Bug: 1602643
# File Overlap Count: 1
# Description:
#   Bug 1602643 - Disable WebRender compositor dinamically for async screenshot r=gw
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D59658
# ==============================================================================

diff -r 0c5fb3a8a94d -r f3b490c076d6 gfx/webrender_bindings/DCLayerTree.cpp
--- a/gfx/webrender_bindings/DCLayerTree.cpp	Thu Jan 23 15:35:45 2020 +0200
+++ b/gfx/webrender_bindings/DCLayerTree.cpp	Thu Jan 23 04:37:28 2020 +0000
@@ -52,9 +52,12 @@
       mDebugCounter(false),
       mDebugVisualRedrawRegions(false),
       mEGLImage(EGL_NO_IMAGE),
-      mColorRBO(0) {}
+      mColorRBO(0),
+      mPendingCommit(false) {}
 
-DCLayerTree::~DCLayerTree() {
+DCLayerTree::~DCLayerTree() { ReleaseNativeCompositorResources(); }
+
+void DCLayerTree::ReleaseNativeCompositorResources() {
   const auto gl = GetGLContext();
 
   DestroyEGLSurface();
@@ -120,7 +123,7 @@
   // Default SwapChain's visual does not need linear interporation.
   mDefaultSwapChainVisual->SetBitmapInterpolationMode(
       DCOMPOSITION_BITMAP_INTERPOLATION_MODE_NEAREST_NEIGHBOR);
-  mCompositionDevice->Commit();
+  mPendingCommit = true;
 }
 
 void DCLayerTree::MaybeUpdateDebug() {
@@ -128,14 +131,30 @@
   updated |= MaybeUpdateDebugCounter();
   updated |= MaybeUpdateDebugVisualRedrawRegions();
   if (updated) {
-    mCompositionDevice->Commit();
+    mPendingCommit = true;
   }
 }
 
+void DCLayerTree::MaybeCommit() {
+  if (!mPendingCommit) {
+    return;
+  }
+  mCompositionDevice->Commit();
+}
+
 void DCLayerTree::WaitForCommitCompletion() {
   mCompositionDevice->WaitForCommitCompletion();
 }
 
+void DCLayerTree::DisableNativeCompositor() {
+  MOZ_ASSERT(mCurrentSurface.isNothing());
+  MOZ_ASSERT(mCurrentLayers.empty());
+
+  ReleaseNativeCompositorResources();
+  mPrevLayers.clear();
+  mRootVisual->RemoveAllVisuals();
+}
+
 bool DCLayerTree::MaybeUpdateDebugCounter() {
   bool debugCounter = StaticPrefs::gfx_webrender_debug_dcomp_counter();
   if (mDebugCounter == debugCounter) {